storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java

```diff
@@ -0,0 +1,2250 @@
+                MetadataVersion.latestTesting()
+        );
+
+        LogValidator.ValidationResult validatedResults = validator.validateMessagesAndAssignOffsets(
+                PrimitiveRef.ofLong(0),
+                metricsRecorder,
+                RequestLocal.withThreadConfinedCaching().bufferSupplier()
+        );
+
+        MemoryRecords validatedRecords = validatedResults.validatedRecords;
+
+        int i = 0;
+        for (RecordBatch batch : validatedRecords.batches()) {
+            assertTrue(batch.isValid());
+            assertEquals(batch.timestampType(), TimestampType.CREATE_TIME);
+            maybeCheckBaseTimestamp(timestampSeq.get(0), batch);
+            assertEquals(batch.maxTimestamp(), batch.maxTimestamp());
+            assertEquals(producerEpoch, batch.producerEpoch());
+            assertEquals(producerId, batch.producerId());
+            assertEquals(baseSequence, batch.baseSequence());
+            assertEquals(partitionLeaderEpoch, batch.partitionLeaderEpoch());
+
+            for (Record record : batch) {
+                record.ensureValid();
+                assertEquals((long) timestampSeq.get(i), record.timestamp());
+                i++;
+            }
+        }
+
+        assertEquals(now + 1, validatedResults.maxTimestampMs, "Max timestamp should be " + (now + 1));
+
+        int expectedShallowOffsetOfMaxTimestamp = 2;
+        assertEquals(expectedShallowOffsetOfMaxTimestamp, validatedResults.shallowOffsetOfMaxTimestamp, "Shallow offset of max timestamp should be 2");
+        assertFalse(validatedResults.messageSizeMaybeChanged, "Message size should not have been changed");
+
+        verifyRecordValidationStats(validatedResults.recordValidationStats, 0, records, true);
+    }
+
+    private MemoryRecords createRecords(List<byte[]> records,
+                                        byte magicValue,
+                                        long timestamp,
+                                        Compression codec) {
+        ByteBuffer buf = ByteBuffer.allocate(512);
+        MemoryRecordsBuilder builder = MemoryRecords.builder(buf, magicValue, codec, TimestampType.CREATE_TIME, 0L);
+
+        AtomicInteger offset = new AtomicInteger(0);
+        records.forEach(item ->
+                builder.appendWithOffset(offset.getAndIncrement(), timestamp, null, item));
+        return builder.build();
+    }
+
+    @Test
+    void testLogAppendTimeWithRecompressionV1() {
+        checkLogAppendTimeWithRecompression(RecordBatch.MAGIC_VALUE_V1);
+    }
+
+    @Test
+    void testLogAppendTimeWithRecompressionV2() {
+        checkLogAppendTimeWithRecompression(RecordBatch.MAGIC_VALUE_V2);
+    }
+
+    @Test
+    void testLogAppendTimeWithoutRecompressionV1() {
+        checkLogAppendTimeWithoutRecompression(RecordBatch.MAGIC_VALUE_V1);
+    }
+
+    @Test
+    void testCompressedV2() {
+        checkCompressed(RecordBatch.MAGIC_VALUE_V2);
+    }
+    @Test
+    void testInvalidOffsetRangeAndRecordCount() {
```

Comment:
Could you please separate the different cases? for example:
```java
    @Test
    void testRecordBatchWithCountOverrides() {
        // The batch to be written contains 3 records, so the correct lastOffsetDelta is 2
        validateRecordBatchWithCountOverrides(2, 3);
    }

    @ParameterizedTest
    @CsvSource({"0,3", "15,3", "-3,3"})
    void testInconsistentCountAndOffset(int lastOffsetDelta, int count) {
        // Count and offset range are inconsistent or invalid
        assertInvalidBatchCountOverrides(lastOffsetDelta, count);
    }

    @ParameterizedTest
    @CsvSource({"5,6", "1,2"})
    void testUnmatchedNumberOfRecords(int lastOffsetDelta, int count) {
        // Count and offset range are consistent, but do not match the actual number of records
        assertInvalidBatchCountOverrides(lastOffsetDelta, count);
    }
```

Reply:
Done.